File: packages/restapi/src/lib/chat/start.ts
import axios from 'axios';
import { getAPIBaseUrls } from '../helpers';
import Constants from '../constants';
import { ChatOptionsType, MessageWithCID } from '../types';
import { IPGPHelper, ISendMessagePayload, PGPHelper, sendMessagePayloadCore, sign } from './helpers';
import * as CryptoJS from 'crypto-js';

export const start = async (
  options: Omit<ChatOptionsType, 'account'>
): Promise<MessageWithCID> => {
  return await startCore(options, PGPHelper);
};

export const startCore = async (
  options: Omit<ChatOptionsType, 'account'>,
  pgpHelper: IPGPHelper,
): Promise<MessageWithCID> => {
  const {
    messageContent = '',
    messageType = 'Text',
    receiverAddress,
    connectedUser,
    apiKey = '',
    env = Constants.ENV.PROD,
  } = options || {};

  const API_BASE_URL = getAPIBaseUrls(env);
  const apiEndpoint = `${API_BASE_URL}/v1/chat/request`;
  const headers = {
    authorization: `Bearer ${apiKey}`,
  };
  const body: ISendMessagePayload = await sendMessagePayloadCore(
    receiverAddress,
    connectedUser,
    messageContent,
    messageType,
    env,
    pgpHelper
  );

  const bodyToBeHashed = {
    fromDID: body.fromDID,
    toDID: body.toDID,
    messageContent: body.messageContent,
    messageType: messageType,
  };

  const hash = CryptoJS.SHA256(JSON.stringify(bodyToBeHashed)).toString();
  const signature: string = await pgpHelper.sign({
    message: hash,
    signingKey: connectedUser.privateKey!,
  });
  const sigType = 'pgp';

  const verificationProof: string = sigType + ':' + signature;
  body.verificationProof = verificationProof;

  return axios
    .post(apiEndpoint, body, { headers })
    .then((response) => {
      return response.data;
    })
    .catch((err) => {
      throw new Error(err);
    });
};

File: packages/push-api-dart/lib/src/chat/src/start.dart
import 'package:push_api_dart/push_api_dart.dart';

Future<MessageWithCID?> start({
  String messageContent = '',
  String messageType = 'Text',
  required String receiverAddress,
  ConnectedUser? connectedUser,
  apiKey = '',
}) async {
  // connectedUser ??= getCachedUser();
  if (connectedUser == null) {
    throw Exception('Account address is required.');
  }

  final SendMessagePayload body = await sendMessagePayload(
    receiverAddress: receiverAddress,
    senderCreatedUser: connectedUser,
    messageContent: messageContent,
    messageType: messageType,
  );

  final bodyToBeHashed = {
    'fromDID': body.fromDID,
    'toDID': body.toDID,
    'messageContent': body.messageContent,
    'messageType': messageType,
  };

  final hash = generateHash(bodyToBeHashed);

  //TODO add sign function parameter values
  final String signature = await sign(
    message: hash,
    privateKey: connectedUser.user.encryptedPrivateKey!,
    publicKey: connectedUser.user.publicKey!,
  );

  const sigType = 'pgp';
  final String verificationProof = '$sigType:$signature';

  body.verificationProof = verificationProof;

  final result = await http.post(
    authorization: apiKey,
    path: '/v1/chat/request',
    data: body.toJson(),
  );

  if (result == null) {
    return null;
  }
  return MessageWithCID.fromJson(result);
}
